### PR TITLE
[5.9][CMake] Support Macros in Linux

### PR DESCRIPTION
**Pull request by rintaro**
Cherry-pick #66199 and #68082 into `release/5.9`

* **Explanation**: Enable Macros in Linux. Swift parser integration is required for enabling macros. Since Linux doesn't have ABI stable `swiftCore` in `/usr/lib` or any other stable location, executables and shared libraries that contains Swift module need to have `RUNPATH` to the runtime in its own toolchain. This PR forces "hosttools" bootstrapping mode, but set the `RUNPATH` to the currently-being-built toolchain.
* **Scope**: CMake build script
* **Risk**: Mid. This PR is carefully crafted not to affect Darwin platforms. For Linux, this is necessary for enabling macros.
* **Testing**: Passes current test suite.
* **Issues**: rdar://104346187
* **Reviewers**: TBA

### CODE REVIEW

**Comment by rintaro** [comment]
https://github.com/apple/llvm-project/pull/7345
https://github.com/apple/swift-syntax/pull/2114
@swift-ci Please Build Toolchain

File: cmake/modules/SwiftUtils.cmake [hunk 1]
```diff
@@ -107,6 +107,16 @@ function(get_bootstrapping_swift_lib_dir bs_lib_dir bootstrapping)
     elseif("${bootstrapping}" STREQUAL "")
       get_bootstrapping_path(bs_lib_dir ${lib_dir} "1")
     endif()
+  elseif(BOOTSTRAPPING_MODE STREQUAL "HOSTTOOLS")
+    if(SWIFT_HOST_VARIANT_SDK MATCHES "LINUX|ANDROID|OPENBSD|FREEBSD")
```
Comment (bnbarham) [inline cmake/modules/SwiftUtils.cmake:111]: There's a bunch of places we're only checking for Linux and others where we check for DARWIN and then else. Given we haven't checked the other platforms, should we just check for LINUX for now?

File: lib/SILOptimizer/Utils/BasicBlockOptUtils.cpp [hunk 2]
```diff
@@ -367,3 +367,12 @@ StaticInitCloner::clone(SingleValueInstruction *initVal) {
   }
   return cast<SingleValueInstruction>(getMappedValue(initVal));
 }
+
+// START: Workaround for C++ interop
```
Comment (bnbarham) [inline lib/SILOptimizer/Utils/BasicBlockOptUtils.cpp:371]: Did you pick a random file or is this the one I randomly picked? Would probably be better to put this in its own file, but given it's only needed for 5.9 ... 🤔

**Review by bnbarham — APPROVED** [review]

**Comment by rintaro** [comment]
https://github.com/apple/llvm-project/pull/7345
https://github.com/apple/swift-syntax/pull/2114
@swift-ci Please test

**Comment by rintaro** [comment]
https://github.com/apple/llvm-project/pull/7345
https://github.com/apple/swift-syntax/pull/2114
@swift-ci Please Build Toolchain

**Comment by rintaro** [comment]
https://github.com/apple/llvm-project/pull/7345
https://github.com/apple/swift-syntax/pull/2114
@swift-ci Please test

**Comment by rintaro** [comment]
https://github.com/apple/llvm-project/pull/7345
https://github.com/apple/swift-syntax/pull/2114
@swift-ci Please Build Toolchain

**Comment by rintaro** [comment]
https://github.com/apple/llvm-project/pull/7345
https://github.com/apple/swift-syntax/pull/2114
@swift-ci Please test

**Comment by rintaro** [comment]
https://github.com/apple/llvm-project/pull/7345
https://github.com/apple/swift-syntax/pull/2114
@swift-ci Please Build Toolchain

**Pull request merged**